src/main/java/it/gov/pagopa/reward/notification/service/iban/outcome/recovery/CompletedKoDiscardedRewardNotificationServiceImpl.java

```diff
@@ -0,0 +1,129 @@
+        return PerformanceLogger.logTimingOnNext("REWARD_NOTIFICATION_IBAN_OUTCOME",
+                rewardsNotificationRepository.findByUserIdAndInitiativeIdAndStatusAndRemedialIdNull(
+                                rewardIban.getUserId(),
+                                rewardIban.getInitiativeId(),
+                                RewardNotificationStatus.COMPLETED_KO)
+                        .flatMap(this::createRemedialNotification)
+                        .collectList(),
+                recovered -> "Recovered %d COMPLETED_KO rewardNotification on userId %s and initiativeId %s".formatted(
+                        recovered.size(),
+                        rewardIban.getUserId(),
+                        rewardIban.getInitiativeId()));
+    }
+
+    private Mono<RewardsNotification> createRemedialNotification(RewardsNotification discarded) {
+        log.info("[REWARD_NOTIFICATION_IBAN_OUTCOME] Found discarded COMPLETED_KO rewardNotification having id {} on userId {} and initiativeId {}",
+                discarded.getId(), discarded.getUserId(), discarded.getInitiativeId());
+
+        return Mono.just(discarded)
+                .flatMap(this::buildRemedialNotification)
+                .flatMap(remedialNotification -> updateDiscardedAndStoreRemedial(discarded, remedialNotification))
+
+                .onErrorResume(e -> {
+                    log.error("[REWARD_NOTIFICATION_IBAN_OUTCOME] Something went wrong while recovering COMPLETED_KO rewardNotification having id {} related to userId {} and initiativeId {}",
+                            discarded.getId(), discarded.getUserId(), discarded.getInitiativeId(), e);
+                    return Mono.empty();
+                });
+    }
+
+    private Mono<RewardsNotification> buildRemedialNotification(RewardsNotification discarded) {
+
+        RewardsNotification remedial = discarded.toBuilder()
+                .status(RewardNotificationStatus.TO_SEND)
+                .exportId(null)
+                .exportDate(null)
+                .iban(null)
+                .rejectionReason(null)
+                .resultCode(null)
+                .feedbackDate(null)
+                .feedbackHistory(Collections.emptyList())
+                .cro(null)
+                .executionDate(null)
+                .remedialId(null)
+                .build();
+
+        setNewIdAndOrdinaryId(remedial, discarded);
+
+
+        return setRemedialNotificationDate(discarded.getInitiativeId(), remedial);
+    }
+
+    private void setNewIdAndOrdinaryId(RewardsNotification out, RewardsNotification input) {
+        String id = input.getId();
+
+        // if recovering a remedial it will calculate the next progressive
+        if (input.getOrdinaryId() != null) {
+            int nextRecoveryProgressiveId = getNextRecoveryProgressiveId(id);
+            out.setId(RECOVERED_ID_PLACEHOLDERS.formatted(input.getOrdinaryId(), RECOVERY_ID_SUFFIX, nextRecoveryProgressiveId));
```

Comment:
if you want to remark the common format logic, maybe it's better to move also the "formatted" call itself to a centralized method thant just the pattern

Reply:
ps: also the nextProgressive logic could be moved inside the same centralized method